### PR TITLE
feat(avatar-group): Avatar group component

### PR DESCRIPTION
**Pull request by 321gillian**
- Simplified version with dedicated group item avatar implementation. 
- Using menu pattern as per - https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/menu_role
- Keyboard nav based off of the menu example: https://www.w3.org/WAI/ARIA/apg/patterns/menubar/examples/menubar-navigation/

### CODE REVIEW

**Comment by jordanstith15** [comment]
@321gillian What is the intended function of a click or enter on one of the circles?

I've noticed that clicking on any of the circles seems to send the exact same data (`detail` is always 1) and hitting enter when focused doesn't have the same data as a click (missing `detail`) according to the doc notification

**Comment by jordanstith15** [comment]
> @321gillian What is the intended function of a click or enter on one of the circles?
> 
> I've noticed that clicking on any of the circles seems to send the exact same data (`detail` is always 1) and hitting enter when focused doesn't have the same data as a click (missing `detail`) according to the doc notification

I'm still seeing the same thing.. maybe the new demo didn't publish yet? I can show you what I mean if that would help

**Comment by 321gillian** [comment]
> > @321gillian What is the intended function of a click or enter on one of the circles?
> > I've noticed that clicking on any of the circles seems to send the exact same data (`detail` is always 1) and hitting enter when focused doesn't have the same data as a click (missing `detail`) according to the doc notification
> 
> I'm still seeing the same thing.. maybe the new demo didn't publish yet? I can show you what I mean if that would help

Having an issue with the build in Jenkins. Will revert once I fix it.

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.scss [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  inline-size: var(--gse-ui-avatar-small-content-size);
+  block-size: var(--gse-ui-avatar-small-content-size);
+  padding: 0;
+  margin: 0;
+  margin-inline-end: var(--gse-ui-avatar-groupSet-gap);
+  overflow: hidden;
+  line-height: 0px;
```
Comment (gavin-everett-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.scss:27]: I dont think this is needed  🤔

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.scss [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+  @include mask;
+
+  &:focus-visible,
+  &:hover {
+    position: relative;
+    z-index: var(--gse-semantic-zIndex-showFocus);
+    mask: none;
+
+    @include focus.gux-focus-ring;
```
Comment (gavin-everett-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.scss:43]: Should this styling just be for focus and not the hover styling ? In Figma it shows the tooltip but without the focus styling.

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.scss [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+  &.gux-last-item {
+    mask: none;
+  }
+
+  .gux-avatar-initials {
+    font-family: var(--gse-ui-avatar-small-initials-fontFamily);
+    font-size: var(--gse-ui-avatar-small-initials-fontSize);
+    font-weight: var(--gse-ui-avatar-small-initials-fontWeight);
+    line-height: 1;
```
Comment (gavin-everett-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.scss:54]: Token need here

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+  @Element()
+  root: HTMLElement;
+
+  /**
+   * Name which is shown as initials. Should be formatted 'Lastname Firstname' for JA, zhCN and KO names.
+   * Names without blank space will show first 2 characters of string.
+   */
+  @Prop()
+  name!: string;
```
Comment (gavin-everett-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx:40]: Is the `!` needed here since you are logging a warning below anyways ?

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+  index: number = 0;
+
+  /**
+   * Removes mask from avatar display if is last in group
+   */
+  @State()
+  lastItemInGroup: boolean = false;
+
+  @State()
+  taxIndexVal: number = -1;
```
Comment (gavin-everett-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx:61]: Is this being used ?

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+  }
+
+  private getAccent(): string {
+    // get current item index in parent element
+
+    if (this.accent !== 'auto') {
+      return this.accent;
+    }
+
+    return (this.index % 12).toString();
```
Comment (gavin-everett-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx:78]: optional change `return this.accent === 'auto'  ? (this.index % 12).toString() : this.accent

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    if (avatarImage && !avatarImage.getAttribute('alt')) {
+      logWarn(this.root, 'Alt attribute is required for slotted image.');
+    }
+  }
+
+  private getDescriptionText(): string {
+    return `${this.name}`;
+  }
```
Comment (gavin-everett-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx:94]: optional but maybe `getFullName()`

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    this.taxIndexVal = 0;
+    resetFocusableSiblings(this.root);
+  }
+
+  render(): JSX.Element {
+    return (
+      <button
+        type="button"
+        role="menuitem"
+        aria-haspopup="false"
```
Comment (gavin-everett-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx:145]: I think this only needs to be specified if the menu-item can open or close a popup? Or will this component also be used from the menu dropdown aswell ?

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group.tsx [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import { Component, h, JSX, Element, Host } from '@stencil/core';
+import { trackComponent } from '@utils/tracking/usage';
+import { logWarn } from '@utils/error/log-error';
+
```
Comment (gavin-everett-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group.tsx:4]: To document default slot here.

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/readme.md [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+<!-- Auto Generated Below -->
+
+
+## Properties
+
+| Property     | Attribute     | Description | Type                              | Default |
+| ------------ | ------------- | ----------- | --------------------------------- | ------- |
+| `maxVisible` | `max-visible` |             | `1 \| 2 \| 3 \| 4 \| 5 \| 6 \| 7` | `7`     |
+
```
Comment (gavin-everett-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar-group/readme.md:13]: I think the readme needs to be updated.

**Comment by gavin-everett-genesys** [comment]
Overall looks good to me 👍 

Some small issue found when manually testing,

1. If I keyboard focus on an avatar and it shows the tooltip and I mouse click on another avatar the keyboard focussed avatar does not get deselected and the tooltip continues showing,

Screenshot 2024-11-18 at 10 52 07

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx (hunk 5 above)
Comment (321gillian) [reply packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx:61]: Yes, it's used to set the initial tab index and the index when clicked.

**Comment by 321gillian** [comment]
> Overall looks good to me 👍
> 
> Some small issue found when manually testing,
> 
> 1. If I keyboard focus on an avatar and it shows the tooltip and I mouse click on another avatar the keyboard focussed avatar does not get deselected and the tooltip continues showing,
> 
> Screenshot 2024-11-18 at 10 52 07

updated to fix this

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx (hunk 5 above)
Comment (gavin-everett-genesys) [reply packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx:65]: minor typo on it `tabIndexVal`

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  @Listen('keydown')
+  onKeydown(event: KeyboardEvent): void {
+    groupKeyboardNavigation(event, this.root);
+
+    switch (event.key) {
+      case 'ArrowRight':
+      case 'ArrowLeft':
+        event.stopPropagation();
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group-item/gux-avatar-group-item.tsx:114]: why is this here and not in `groupKeyboardNavigation`?

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group.tsx [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    if (focusedChildren.length > 0) {
+      focusedChildren.forEach((child: HTMLGuxAvatarGroupItemBetaElement) => {
+        if (child !== target) {
+          child.hideTooltip();
+        }
+      });
+    }
+  }
+
+  private handleClick(event: MouseEvent) {
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group.tsx:66]: Why is this needed? is the clicked button not focused automatically?

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group.tsx [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    if (groupItems.length > 0) {
+      const firstGroupItem = groupItems[0] as Element;
+      focusGroupItem(firstGroupItem);
+    }
+  }
+
+  private getRootChildren(root: HTMLElement): Element[] {
+    const slot = root.querySelector('slot');
+
+    return slot ? slot.assignedElements() : Array.from(root.children);
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group.tsx:45]: why is this so complicated? Have a look at `gux-segmented-control-beta` for a simple  way to do this

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group.tsx [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+  @Listen('mouseover')
+  onMouseOver(event: MouseEvent): void {
+    const target = event.target as HTMLElement;
+    const children = Array.from(
+      this.root.querySelectorAll('gux-avatar-group-item-beta')
+    ) as HTMLGuxAvatarGroupItemBetaElement[];
+    const focusedChildren = children.filter(child =>
+      child.matches(':focus-within')
+    );
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group.tsx:56]: if there can only ever be one child with focus within you can use `find` instead of `filter`

File: packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group.tsx (hunk 12 above)
Comment (321gillian) [reply packages/genesys-spark-components/src/components/beta/gux-avatar-group/gux-avatar-group.tsx:65]: Changed it around a bit to make it clearer

**Review by daragh-king-genesys — APPROVED** [review]

**Review by jordanstith15 — CHANGES_REQUESTED** [review]
Hitting "enter" to click is causing a console error. Clicking with mouse does not

**Review by jordanstith15 — APPROVED** [review]

**Review by jordanstith15 — APPROVED** [review]

**Review by daragh-king-genesys — APPROVED** [review]
